Index: linux-4.14.131/arch/mips/Kbuild.platforms
===================================================================
--- linux-4.14.131.orig/arch/mips/Kbuild.platforms
+++ linux-4.14.131/arch/mips/Kbuild.platforms
@@ -27,6 +27,7 @@ platforms += pmcs-msp71xx
 platforms += pnx833x
 platforms += ralink
 platforms += rb532
+platforms += realtek
 platforms += sgi-ip22
 platforms += sgi-ip27
 platforms += sgi-ip32
Index: linux-4.14.131/arch/mips/Kconfig
===================================================================
--- linux-4.14.131.orig/arch/mips/Kconfig
+++ linux-4.14.131/arch/mips/Kconfig
@@ -630,6 +630,21 @@ config RALINK
 	select ARCH_HAS_RESET_CONTROLLER
 	select RESET_CONTROLLER
 
+config REALTEK
+	bool "Realtek RLX based SoC support"
+	select BOOT_RAW
+	select DMA_NONCOHERENT
+	select IRQ_MIPS_CPU
+	select USE_OF
+	select SYS_SUPPORTS_MIPS16
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_HAS_EARLY_PRINTK
+	select COMMON_CLK
+	select TIMER_OF
+	select CLKSRC_MMIO
+	help
+	 Support for Realtek RLX SoC based boards
+
 config SGI_IP22
 	bool "SGI IP22 (Indy/Indigo2)"
 	select FW_ARC
@@ -1032,6 +1048,7 @@ source "arch/mips/pic32/Kconfig"
 source "arch/mips/pistachio/Kconfig"
 source "arch/mips/pmcs-msp71xx/Kconfig"
 source "arch/mips/ralink/Kconfig"
+source "arch/mips/realtek/Kconfig"
 source "arch/mips/sgi-ip27/Kconfig"
 source "arch/mips/sibyte/Kconfig"
 source "arch/mips/txx9/Kconfig"
Index: linux-4.14.131/arch/mips/boot/dts/Makefile
===================================================================
--- linux-4.14.131.orig/arch/mips/boot/dts/Makefile
+++ linux-4.14.131/arch/mips/boot/dts/Makefile
@@ -10,6 +10,7 @@ dts-dirs	+= ni
 dts-dirs	+= pic32
 dts-dirs	+= qca
 dts-dirs	+= ralink
+dts-dirs	+= realtek
 dts-dirs	+= xilfpga
 
 obj-y		:= $(addsuffix /, $(dts-dirs))
