--- a/drivers/spi/Kconfig	
+++ b/drivers/spi/Kconfig	
@@ -527,6 +527,11 @@
 config SPI_PXA2XX_PCI
 	def_tristate SPI_PXA2XX && PCI && COMMON_CLK
 
+config SPI_REALTEK
+	tristate "Realtek RTL819x SPI controller driver"
+	help
+	  This selects a driver for Realtek RTL8196E and RTL8197D SoC
+ 
 config SPI_ROCKCHIP
 	tristate "Rockchip SPI controller driver"
 	help
@@ -619,6 +624,11 @@
 	help
 	  SPI driver for SuperH HSPI blocks.
 
+config SPI_SHEIPA
+	tristate "Sheipa SPI controller"
+	help
+	  SPI driver for Sheipa Controller 
+
 config SPI_SIRF
 	tristate "CSR SiRFprimaII SPI controller"
 	depends on SIRF_DMA
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile	
@@ -109,6 +109,9 @@
 obj-$(CONFIG_SPI_XTENSA_XTFPGA)		+= spi-xtensa-xtfpga.o
 obj-$(CONFIG_SPI_ZYNQMP_GQSPI)		+= spi-zynqmp-gqspi.o
 
+obj-$(CONFIG_SPI_REALTEK)			+= spi-realtek.o
+obj-$(CONFIG_SPI_SHEIPA)			+= spi-sheipa.o
+
 # SPI slave protocol handlers
 obj-$(CONFIG_SPI_SLAVE_TIME)		+= spi-slave-time.o
 obj-$(CONFIG_SPI_SLAVE_SYSTEM_CONTROL)	+= spi-slave-system-control.o